### ISSUE TITLE
Proper API for working with frames

### ISSUE THREAD

**Issue by abotalov** [teamcapybara/capybara#1365]
Hi,

I'd like to contribute a new API for working with frames (it can be done for e.g. 2.5):
- `Session#switch_to_frame(:top_level)`
- `Session#switch_to_frame(:parent)` (switch_to.parent_frame will be added in Selenium-Webdriver 2.43)
- `Session#switch_to_frame(element)` (element is Capybara::Node::Element)
- `Session#within_frame(element)`

Old `within_frame(frame_handle)` where "frame_handle" is passed to the driver (and in reality can be not only handle but also something else) will be deprecated like it was done with `within_window`.

Please approve.

**Comment by jtrupiano** [comment]
Did this new API ever get merged in?  I need the #switch_to_frame methods.

**Comment by route** [comment]
@jtrupiano I don't think so. Ended up here as proposal

**Comment by jtrupiano** [comment]
@route thanks for the reply.  A question for you (and the lurkers) regarding the following situation.  I am directly working with Session instances (no naked calls to #within_frame, etc).  I have scenarios where I want to observe the same behavior when a request comes directly from the browser vs. when the request is loaded into an iframe on another page.  So in my test code, I want a "handle" of sorts to the top level Session-like object.  In other words, something like this...

    # return an object that behaves like Session that is scoped to an embedded iframe (if necessary)
    def get_root(session)
      if session.has_css?('iframe#MyFrame')
        return session.get_iframe('iframe#MyFrame') # should behave like a Session
      else
        return session
      end
    end
    
    root = get_root(session)
    root.fill_in 'Name', 'John'
    root.submit # etc

This is admittedly stripped down for illustrative purposes, but I want to re-use all of that code after the call to #get_root as if it were unaware that it were happening within an iframe.  So what's the recommended way to get a pointer/handle to the iframe that either is or behaves like a Session?

**Comment by route** [comment]
I'm a little out of context but why you cannot use `within_frame` even if you manage session yourself? So basically it pushes an iframe to the stack and when block ends pops it.

**Comment by jtrupiano** [comment]
I don't understand how to do that.  #within_frame doesn't yield a root-level "Session-like" object for me to work with (or does it??).

    session = Capybara::Session.new(Capybara.current_driver)
    session.visit "http://something.com"
    session.within_frame('iframe#MyFrame') do 
      # how would I invoke anything on that iframe from here?
      fill_in 'Name', 'John' # this will throw an error because I'm managing session myself
    end

What am I missing / not understanding?

**Comment by route** [comment]
I haven't tried and can be mistaken
```
session = Capybara::Session.new(Capybara.current_driver)
session.visit "http://something.com"
session.within_frame('iframe#MyFrame') do 
  # how would I invoke anything on that iframe from here?
  session.fill_in 'Name', 'John'
end
```

**Comment by jtrupiano** [comment]
Oh I get it now.  #within_frame actually alters the state of the variable session.  That's what I was missing.  Thanks for helping me think through this, @route.

btw I'd still be happier if I had #switch_to_frame :)

**Comment by route** [comment]
@jtrupiano No problem ;)

**Comment by twalpole** [comment]
Driver authors please look at PR #1722 and give relevant feedback

**Comment by twalpole** [comment]
Closed via PR #1803
